haskell-language-server.cabal

```diff
@@ -141,6 +141,56 @@ test-suite hls-cabal-fmt-plugin-tests
   if flag(isolateCabalfmtTests)
     build-tool-depends: cabal-fmt:cabal-fmt ^>=0.1.6
 
+-----------------------------
+-- cabal-gild plugin
+-----------------------------
+
+flag cabalgild
+  description: Enable cabal-gild plugin
+  default:     True
+  manual:      True
+
+common cabalgild
+  if flag(cabalgild)
+    build-depends: haskell-language-server:hls-cabal-gild-plugin
+    cpp-options: -Dhls_cabalgild
+
+flag isolateCabalGildTests
+  description: Should tests search for 'cabal-gild' on the $PATH or shall we install it via build-tool-depends?
+  -- By default, search on the PATH
+  default:     False
+  manual:      True
+
+library hls-cabal-gild-plugin
+  import:           defaults, pedantic, warnings
+  exposed-modules:  Ide.Plugin.CabalGild
+  hs-source-dirs:   plugins/hls-cabal-gild-plugin/src
+  build-depends:
+    , base            >=4.12 && <5
+    , directory
+    , filepath
+    , ghcide          == 2.7.0.0
+    , hls-plugin-api  == 2.7.0.0
+    , lsp-types
+    , text
+    , mtl
+    , process-extras
+
+test-suite hls-cabal-gild-plugin-tests
+  import:           defaults, pedantic, test-defaults, warnings
+  type:             exitcode-stdio-1.0
+  hs-source-dirs:   plugins/hls-cabal-gild-plugin/test
+  main-is:          Main.hs
+  build-depends:
+    , base
+    , directory
+    , filepath
+    , haskell-language-server:hls-cabal-gild-plugin
+    , hls-test-utils        == 2.7.0.0
+
+  if flag(isolateCabalGildTests)
+    build-tool-depends: cabal-gild:cabal-gild ^>=1.1
```

Comment:
Do we need to do it like `cabal-fmt`? Or can we do it more like the other formatters, where we use the library by default and have an option to use an executable optionally? I thought we did it like this because `cabal-fmt` was particularly difficult to build? But maybe `cabal-gild` is easier?

Reply:
`cabal-gild` is easier, but it doesn't compile with ghc 9.2. We can disable the plugin for that GHC version, I don't feel strongly.

Reply:
I don't think it would be terribly hard to add support for GHC 9.2. Let me see what I can do.

Aside from that, Gild doesn't really expose a high-level "format" function. That's because I primarily intended for it to be an application rather than a library. The [`mainWith`](https://hackage.haskell.org/package/cabal-gild-1.1.0.0/docs/CabalGild-Main.html#v:mainWith) function can be used that way, but it's not very convenient. I could add a `format` function. What should the type signature be? Something like `ByteString -> IO ByteString`?

Reply:
A `format` function would be nice, but we need to be able to pass the `FilePath` option for `cabal-gild: discover` feature to work.
Perhaps a function `Opts -> ByteString -> IO ByteString`?

Reply:
> That's because I primarily intended for it to be an application rather than a library.

That's a good reason for us to interact with it that way. 

It can be nice to at least have the option to use a compiled-in library version instead of shelling out to a process, but that does lead to the dread https://github.com/haskell/haskell-language-server/issues/411

Reply:
I added support for GHC 9.2 in Gild version 1.1.1.0: https://github.com/tfausak/cabal-gild/releases/tag/1.1.1.0

I'll noodle on a high level formatting function. The existing `mainWith` function can be used in a pure manner, it's just kind of annoying. See the test suite for an example: https://github.com/tfausak/cabal-gild/blob/747550fb20c4ab4003a199ebacd07130ce16796c/source/test-suite/Main.hs#L999